server/commands/connected_objects.go

```diff
@@ -0,0 +1,355 @@
+
+	// Limit limits the results yielded by the ConnectedObjects API.
+	Limit uint32
+}
+
+type ConnectedObjectsCommand struct {
+	Datastore        storage.OpenFGADatastore
+	Typesystem       *typesystem.TypeSystem
+	ResolveNodeLimit uint32
+}
+
+func (c *ConnectedObjectsCommand) streamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	depth, ok := graph.ResolutionDepthFromContext(ctx)
+	if !ok {
+		ctx = graph.ContextWithResolutionDepth(ctx, c.ResolveNodeLimit)
+	} else {
+		if depth == 0 {
+			return serverErrors.AuthorizationModelResolutionTooComplex
+		}
+
+		ctx = graph.ContextWithResolutionDepth(ctx, depth-1)
+	}
+
+	storeID := req.StoreID
+
+	targetUserType, _ := tuple.SplitObject(req.User.Object)
+
+	targetUserRef := &openfgapb.RelationReference{
+		Type:     targetUserType,
+		Relation: req.User.Relation,
+	}
+
+	sourceObjRef := &openfgapb.RelationReference{
+		Type:     req.ObjectType,
+		Relation: req.Relation,
+	}
+
+	// build the graph of possible edges between object types in the graph based on the authz model's type info
+	g := graph.BuildConnectedObjectGraph(c.Typesystem)
+
+	// find the possible incoming edges (ingresses) between the target user reference and the source (object, relation) reference
+	ingresses, err := g.RelationshipIngresses(sourceObjRef, targetUserRef)
+	if err != nil {
+		return err
+	}
+
+	for _, ingress := range ingresses {
+		switch ingress.Type {
+		case graph.DirectIngress:
+			err := c.reverseExpandDirect(ctx, &reverseExpandRequest{
+				storeID:          storeID,
+				ingress:          ingress,
+				sourceObjectRef:  sourceObjRef,
+				targetUserRef:    req.User,
+				contextualTuples: req.ContextualTuples,
+				limit:            req.Limit,
+			}, resultChan, foundObjectsMap, foundCount)
+			if err != nil {
+				return err
+			}
+
+		case graph.TupleToUsersetIngress:
+			err := c.reverseExpandTupleToUserset(ctx, &reverseExpandRequest{
+				storeID:          storeID,
+				ingress:          ingress,
+				sourceObjectRef:  sourceObjRef,
+				targetUserRef:    req.User,
+				contextualTuples: req.ContextualTuples,
+				limit:            req.Limit,
+			}, resultChan, foundObjectsMap, foundCount)
+			if err != nil {
+				return err
+			}
+
+		default:
+			return errors.Errorf("unsupported ingress type")
+		}
+	}
+
+	return nil
+}
+
+// StreamedConnectedObjects yields all of the objects of the provided objectType that
+// the given user has a specific relation with. The results will be limited by the request
+// limit. If a 0 limit is provided then all objects of the provided objectType will be
+// returned.
+func (c *ConnectedObjectsCommand) StreamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string, // object string (e.g. document:1)
+) error {
+
+	var foundCount *uint32
+	if req.Limit > 0 {
+		foundCount = new(uint32)
+	}
```

Comment:
This looks weird because then you have to check `if foundCount != nil`

Reply:
This was less confusing to me than the alternative where `foundCount == 0`. What if foundCount is legitimately 0 instead of just the zero value?

The reason for the pointer here is that `foundCount` is unused if the limit is unspecified (0 value).

Reply:
> foundCount is unused if the limit is unspecified

Can't we count regardless of whether the limit was set or not?

Reply:
Sure, but then what do we do in the check
```
if foundCount > X {...}
```
what should `X` be? I could set it to `math.MaxUint64`, but semantically there should be no limit if the limit is omitted. So this is why I chose the path I did. If limit is omitted (default 0 value) then we don't constrain the search to a maximum number of objects.